### ISSUE TITLE
The default protocol should be 'http:'

### ISSUE THREAD

**Issue by rockuw** [jhiesey/stream-http#32]
In order to be consistent with node's http lib. Or the browser will complain like this:

```
TypeError: Failed to execute 'fetch' on 'Window': Failed to parse URL from //10.101.168.94:6000/ (req "error")
```

note `//10.101.168.94:6000/` doesn't have `http:` prefix.

**Comment by jhiesey** [comment]
It's weird that fetch doesn't support protocol relative urls (which start with `//`). This isn't an issue with the XHR case.

The correct solution is going to be to explicitly set the protocol to the protocol of the page, which only needs to happen in the fetch case. I'll get to this in the next few days.

(note that defaulting to http isn't really right, since it will just throw security errors on https pages).

**Comment by rockuw** [comment]
> ...explicitly set the protocol to the protocol of the page

A HTTP request may be a CORS request. So we cannot set it to the page's protocol.

Defaulting to http should be OK. HTTPS should set the protocol to 'https:' explicitly.
From https://nodejs.org/api/http.html :

```
Options:

protocol: Protocol to use. Defaults to 'http'.
```

**Comment by jhiesey** [comment]
I was under the impression that an HTTPS page can't ever make HTTP requests without triggering an error, CORS or no. Is that not the case?

**Comment by jhiesey** [comment]
In the spec (http://www.w3.org/TR/cors/#user-agent-security) it looks like browsers are allowed to choose either behavior. Does Chrome actually allow this? In previous testing I remember it not working.

**Comment by jhiesey** [comment]
I can't reproduce this locally in Chrome, i.e. it works fine for me with
```
http.get('//localhost:8080/index.html', function (res) ...
```
What arguments are you passing to `http.request` or `http.get`? Are you testing this in Chrome or another browser?

When I try
```
http.get('//10.101.168.94:6000/', function (res) ...
```
I actually get a different error, `GET http://localhost:6000/ net::ERR_UNSAFE_PORT`, indicates that the browser is correctly interpreting the url but then deciding not to send the request to port 6000 for security reasons.

(note that I keep assuming Chrome since I don't think any other browsers will pass the feature tests that enable using the fetch API in `stream-http`)

**Comment by jhiesey** [comment]
I tried this in Chrome; when I try to request `http://localhost:8080/` from `https://localhost:8000/` it fails with `Mixed Content: The page at 'https://localhost:8000/' was loaded over HTTPS, but requested an insecure resource 'http://localhost:8080/'. This request has been blocked; the content must be served over HTTPS.`

So I'm unconvinced that defaulting to http is the right choice, but I still want to know why @rockuw was seeing the reported error. That shouldn't be happening.

**Comment by parshap** [comment]
@jhiesey even if it doesn't result in the best user experience, shouldn't this module match the same interface as node's native http module? That would be to default to `http:`. If users want to default to the current page's protocol, they should do that explicitly or use a wrapper http module to do so.

**Comment by jhiesey** [comment]
I don't yet understand the case where adding `http:` fixes @rockuw's issue, since I can't reproduce exactly the same issue.

It would also break the behavior of protocol-relative urls, which *should* be working now but would break if I made that change. Given that I see cases where defaulting to `http:` will fail yet I don't (yet) understand the cases where it improves anything I'd rather leave it as is.

If there aren't any downsides, it seems nice to allow things that usually work in the browser even if they don't in node, since this after all designed to run in the browser. On the other hand if there is a valid case where defaulting to `http:` actually improves things then I agree it's worth changing the behavior.

**Comment by rockuw** [comment]
@jhiesey Here's code (before browerify) to repro the problem:

```javascript
var http = require('http');

window.onload = function () {
  console.log('onload');
  http.get({
    host: 'localhost',
    path: '/hello',
  });
  console.log('done');
};
```

And I got this error:

```
Fetch API cannot load file://localhost/hello. URL scheme must be "http" or "https" for CORS request.
```

While this works in node.js. With this incompatibility, I cannot just port my javascript code to browser.

**Comment by jhiesey** [comment]
Opening a file directly instead of using a server is an interesting edge case that I hasn't considered. This ought to work, I agree. I think defaulting to 'http:' if `global.location.protocol` is neither 'http:' nor 'https:' would be reasonable. (by contrast, in this case it is 'file:'). I can do this soon if you don't get to it first!

Are you loading the page directly from a file just for development or for another reason? I'm just curious.

**Comment by jhiesey** [comment]
And also thanks for finding the bug! I didn't even think of this case.

**Comment by rockuw** [comment]
> Are you loading the page directly from a file just for development or for another reason? I'm just curious.

Just for development.

**Comment by rockuw** [comment]
@jhiesey 

> I can do this soon if you don't get to it first!

Go ahead and thank you. :)

**Comment by jhiesey** [comment]
Fixed in https://github.com/jhiesey/stream-http/commit/c36da3bade632e46b4db25e440f0d295257431cf and published as v2.0.5
